proof/refine/ARM/KHeap_R.thy

```diff
@@ -3743,6 +3743,28 @@ lemma state_refs_of_cross:
       \<Longrightarrow> P (state_refs_of' s')"
   by (clarsimp simp: state_refs_of_cross_eq elim!: rsubst[where P=P])
 
+lemma ct_not_inQ_cross:
+  "\<lbrakk>(s,s') \<in> state_relation; ct_not_in_q s; valid_queues' s'; cur_tcb' s'; pspace_aligned s;
+    pspace_distinct s\<rbrakk>
+   \<Longrightarrow> ct_not_inQ s'"
+  apply (clarsimp simp: ct_not_inQ_def ct_not_in_q_def)
+  apply (prop_tac "scheduler_action s = resume_cur_thread")
+   apply (clarsimp simp: state_relation_def)
+   apply (metis sched_act_relation.simps Structures_A.scheduler_action.exhaust
+                scheduler_action.simps)
+  apply (clarsimp simp: not_queued_def)
+  apply (rule ccontr)
+  apply (prop_tac "obj_at' tcbQueued (ksCurThread s') s'")
+   apply (clarsimp simp: obj_at_simps cur_tcb'_def)
+  apply (frule curthread_relation)
+  apply (frule state_relation_ready_queues_relation)
+  apply (clarsimp simp: valid_queues'_def inQ_def ready_queues_relation_def obj_at_simps)
+  by (metis Structures_H.kernel_object.case(6))
+
+method add_ct_not_inQ =
+  rule_tac Q="\<lambda>s'. ct_not_inQ s'" in corres_cross_add_guard
+  , fastforce intro: ct_not_inQ_cross simp: valid_sched_def
```

Comment:
Seeing that the one for the next PR of this line has another method defined that is almost the same, maybe we can have a “meta” method, something like
```
method add_sch_act_wf P R=
   rule_tac Q=R in corres_cross_add_guard, fastforce intro: R simp: valid_sched_def
   ```
then we can define each instance using this. I am not sure the difference among the instances are, but I’d imagine it is possible to write something that cover most of the cases, if not all.

Also, there probably will be situations where we want to be adding more than one conjunct at a time. Then it will be possible to use this abstracted (parameterised) method to handle multiple additions, instead of applying multiple method sequentially.

Reply:
I think these methods are simple enough that we're not gaining much from doing this.

There is a bit of variation in these methods, though I'm not entirely sure it's necessary that they do vary. One uses `dest` rather than `intro`, another uses `intro!` instead. It seems plausible that we would like to add further things to `simp:` later. Having one method that generates all these means that it's possible that if that method were changed, we could get breakages in many places. The asserts are used all over the place now, so it could be a lot of work to fix that.

If you would like me to try this, then it might be best to merge all these PRs first and then try, so that I can see whether one general method will work

Reply:
Since these crossing PRs follow basically the same pattern, I am just wondering if we could be more systematic with this process. Then things will be easier when we want to repeat this process for a new invariant etc. in the future (or even now).

For example, currently these `add_xxx` methods are defined in various locations (in KHeap_R mostly I think). But thinking ahead, it probably makes sense to collect them in one place. And maybe abstract things away a bit more if possible as I mentioned in the first comment, by having a template method.

I am not sure if breakage etc. mentioned is a real issue here. We are talking about having a template for applying a one particular rule (`corres_cross_add_guard`). The potential variations come from how we write the actual cross lemmas (the parameters). In that sense, I believe it is worthwhile to consider if we can write these cross lemmas in a more uniform way. But as for the template itself, if we have to accomodate something new, we can most likely work out a conservative change or if it has to be the new thing can be dealt with separately, most likely.

And, when we call these methods, the way currently things are done means we might have 3-4 (or potentially more in the future) lines of `add_xxx`, and all of them use the same lemma `corres_cross_add_guard`. So, having a template method that is It would make sense to have a way to say “cross A, B, and C” or “cross all that can be crossed”.